### PR TITLE
Recovery campaign: generated restart scenarios, nightly, with a save/restore benchmark

### PR DESCRIPTION
**Pull request by hhenson**
The test pack for recovery, as asked: variations of code paths and depths; clean start, RECOVER and snapshot recovery; permutations of `map_`, `mesh_`, `dmap_` (in process and across processes), `reduce` and `spawn_` pipelines. Plus the save/restore performance tests.

## The campaign -- `tools/recovery`

A **scenario** is a chain (`dmapp__map__total`: layers over a leaf, to depth 3), a placement (component outermost, or *inside* the worker), a host (main graph, or a `spawn_` stage whose sink is in another process), a mode, a generated presence-aware event stream, and its cuts (single / several / every cycle).

| Mode | What comes back | What must hold |
|---|---|---|
| `clean` | -- | the oracle; a run with nothing to compare is an error |
| `snapshot` | state, from a component checkpoint | **delta-for-delta** equality with the unbroken run |
| `recover` | recorded *inputs*, re-seeded (`RECOVER` mode) | **value** equality for a consumer that starts with the day; stateless graphs |

**Three ways a green campaign can lie, each closed**

- *Comparing nothing to nothing* -- a sample is re-run with the same cuts and **no recovery**; a mode whose controls never differ is `VACUOUS` and fails.
- *Expected refusals rotting* -- `model.expected` names each refusal and why; one that starts to run fails as loudly as the reverse.
- *A defect quietly fixed or spreading* -- a known defect is a **family** (the relation that makes it reachable), not a list of recipes. Members that fail are `known`; a failure outside every family fails; a large family that stops failing is `RETIRED` and fails, so the fix deletes it.

## What it found

A **restored keyed input iterates its keys in a different order** from the unbroken run when there is a removal before a cut and another after it: `(0, 3)` unbroken, `(3, 0)` restored. Only an order-sensitive `reduce` shows it (`76` vs `40`). One of the first eight failures had no `dmap_` or `spawn_` in it, so it predates the recovery stack. Pinned as a strict `xfail` with a three-cycle minimal stream and a control (either removal alone restarts invisibly). **Not fixed here, and the root cause is not established**: the keys are native ints; the order of adds and removals inside one delta is ruled out (all three orderings fail alike); and the C++ counterpart with the same stream **passes** (`test_reduce_checkpoint.cpp`) for a reason I have not found. That difference is where to start.

The campaign also corrected *me*: my first family relation asked for the two removals in the cycles either side of the cut, and the next seed failed one scenario outside it (removal, an ordinary tick, then the cut and a removal). The relation is now "a removal on each side of some cut" -- on that seed 125 of 168 stateful-reduce scenarios are members, all 6 failures are among them, 119 members pass, and the 43 outside pass.

## Where it runs

- **Every test run**: `test_recovery_campaign.py`, every fifth `pr` scenario, ~4 s.
- **Nightly**, `recovery-nightly.yml`: candidate core + persistence wheels -> 16 shards, run number as seed -> a **verdict** job. A shard is too small a sample to call a mode vacuous or a family retired, so shards never fail the run; `merge` judges the totals and that every shard reported, and needs no hgraph.
- **This PR** runs the `pr` profile through those same jobs, so the workflow is not first exercised at night.

## Performance -- `test_recovery_benchmark.cpp` (`[.][recovery-benchmark]`)

`map_` as the bar, then `dmap_` as a member and hosting a component, in process and across 4 processes, at 5k/10k/20k/40k keys. It **requires** cost per key to stay flat (guardrail iv). macOS arm64, µs per key, 5k -> 40k:

| Variant | Save | Restore |
|---|---|---|
| `map_` (the bar) | 1.50 -> 1.72 | 2.70 -> 3.12 |
| `dmap_` member, in process | 1.52 -> 1.87 | 3.21 -> 3.73 |
| `dmap_` hosting, in process | 1.57 -> 1.88 | 3.38 -> 3.74 |
| `dmap_` member, 4 processes | 0.94 -> 0.94 | 2.12 -> 1.93 |
| `dmap_` hosting, 4 processes | 0.75 -> 0.98 | 2.61 -> 2.04 |

Flat everywhere; across processes it beats the single-graph bar. Raw rows for both passes are in `benchmarks/results/recovery-20260919/`; Linux numbers follow in a comment.

## Validation so far

`pr` profile locally: 100/100 (snapshot 82, recover 18), 76 of 83 controls sensitive; sharded run + `merge` on a Python with no hgraph: exit 0, and exit 1 with a shard removed. Nightly profile on Linux (32 shards) and the Linux benchmark: running, results in a comment.

Independent of #998 (the review follow-ups).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by hhenson** [comment]
**Nightly profile on Linux** (real install of this branch, 32 shards, on a private Linux validation host), judged by the same `merge` the workflow uses:

| Seed | Commit | Scenarios | pass | known | fail | recover | controls sensitive | Wall | Verdict |
|---|---|---|---|---|---|---|---|---|---|
| 2 | `c58a1a7aa` | 3,480 | 3,474 | 5 | **1** | 424 / 424 | 691 / 721 | 256 s | **exit 1** |
| 2 | `3acbf9b55` | 3,480 | 3,474 | 6 | 0 | 424 / 424 | 691 / 721 | 254 s | exit 0 |

The first row is the campaign doing its job on me: one failure fell *outside* the known-defect family because my relation was too narrow (see the description). The second row is the same seed with the corrected relation -- 6 failing of 125 members. Coverage of that run: hosts graph 1,952 / spawn 1,528; placement outer 2,800 / inner 680; every layer in 1,780 scenarios; depth 0-3 = 56 / 224 / 896 / 2,304; cuts every 1,181 / several 1,147 / single 1,152. Seed 3 is running.

**Benchmark on Linux** (Threadripper 9980X, GCC 14 Release, 2 passes): the flatness requirement passed on both. Rows and a summary for both machines are in `benchmarks/results/recovery-20260919/`. One thing it shows that I am **not** explaining away: restore in a single process is not flat on Linux -- the `map_` *bar* rises 1.75-1.90x per key over an 8x growth in keys (1.16x on macOS). Nothing is quadratic (8.00 would be), it is the bar rather than `dmap_` that rises most, and the process-hosted variants do not rise at all, which fits a working set outgrowing cache -- but that is an inference from the rows, not a profile. Recorded as open in the results README.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by hhenson** [comment]
**Seed 3, and what it found.** Nightly profile on a private Linux validation host, real install, 32 shards:

| Seed | Commit | Scenarios | pass | known | fail | Verdict |
|---|---|---|---|---|---|---|
| 3 | `c0f1f5aa5` | 3,480 | 3,470 | 9 | **1** | **exit 1** |
| 3 | `fd32824ff` | 3,480 | 3,470 | 10 (9 + 1) | 0 | exit 0 |

The failure was a **RECOVER** scenario in a purely stateless chain (`mesh__mesh__mesh__doubled`), and it is not about recovery. Reduced to two cycles with no recovery anywhere:

```
map_(inner, {0: {}, 5: {1: 1}})      inner = map_(doubled)   ->  {0: {}, 5: {1: 2}}
                                     inner = mesh_(doubled)  ->  {5: {1: 2}}
```

A `mesh_` **started** over an empty key set emits nothing, while one whose keys are all removed **later** keeps its valid empty output -- so its output depends on its history, and RECOVER, being a fresh start with re-seeded inputs, cannot reproduce it. That is already ruled and accepted (parity matrix: *no change means no tick*, "already accepted for `mesh_` over an initially empty key set"), so it is modelled as the consequence of a ruling, **not a defect**: family `mesh-empty-input-no-tick` = mode recover, a `mesh_` layer at depth *k*, an empty collection at depth *k* at some cut. It has exactly one member on each of seeds 2 and 3, and the seed-3 member is the failure. (My first guess -- that RECOVER cannot re-create an emptied inner collection -- did not reproduce under `map_`, which is what pointed at `mesh_`.)

**A parity question this surfaced, not decided here.** Against the pinned 0.5 reference the two operators are *both* different on that input, in opposite directions: the reference gives `map_ -> {5: ...}` and `mesh_ -> {0: {}, 5: ...}`. The `mesh_` side is the accepted ruling. The `map_` side -- emitting an empty map where the reference emits nothing -- sits on the *other* side of the same ruling, like the three temporal operators in #822. I have not checked whether the parity campaign already has a family for it.

Also in this push: the SonarCloud findings (generator split with bit-identical scenarios by sha256 for pr/7, nightly/2 and nightly/3; the shard job now syncs run-time dependencies from the lockfile, wheels only, and installs this run's wheels with `--no-deps`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: .github/workflows/recovery-nightly.yml [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+          name: recovery-candidate-wheel
+          path: dist
+      - uses: actions/download-artifact@37930b1c2abaa49bbe596cd826c3c89aef350131 # v7
+        with:
+          name: recovery-candidate-persistence-wheel
+          path: persistence-dist
+      - name: Install the candidate
+        run: >-
+          python -m pip install --only-binary=:all:
+          dist/*.whl persistence-dist/*.whl "pyarrow==25.0.0"
```
[the remark this answers is not in the document]
Comment (hhenson) [reply .github/workflows/recovery-nightly.yml]: Right about the versions (core `0.0.0`, persistence `hgraph>=0.8.0`), and the install you reviewed would have failed for it. It was already replaced when this landed: the shard job now syncs run-time dependencies from the lockfile (`uv sync --frozen --no-build --no-default-groups --no-install-workspace`) and installs both of this run's wheels with `--no-deps`, as `test-platform-wheel.yml` does, so nothing is resolved against the candidate version (`1d0789420`).

File: tools/recovery/run.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                              extra={"traceback": traceback.format_exc(limit=6)})
+            seconds = time.monotonic() - started
+            if expectation.outcome == "refused":
+                return Result(scenario, "fail",
+                              "expected a refusal (" + expectation.reason + ") and it ran; if a limit was "
+                              "lifted, update tools/recovery/model.py:expected", seconds)
+            if actual != oracle:
+                family = known_defect(scenario)
+                if family is not None:
+                    return Result(scenario, "known", f"{family.id}: " + _first_difference(oracle, actual), seconds)
```
[the remark this answers is not in the document]
Comment (hhenson) [reply tools/recovery/run.py]: Agreed, and I had the same worry about this family's breadth. Fixed in `915f1982a`: membership is a relation over the *recipe* and says nothing about what went wrong, so a mismatch is filed as known only when it has the family's **signature**. For `tsd-restored-slot-order` the same scenario is re-run with the reduction swapped for a probe that reports the keyed input's iteration order, and that has to differ too. For `mesh-empty-input-no-tick`, empty collections are taken out of both runs and nothing else may differ. Otherwise it is a failure that says "in family X, but the mismatch does NOT have its signature". All 16 real instances from two nightly seeds are confirmed by their signatures; a test fabricates mismatches inside member scenarios and they are refused.

File: tools/recovery/cli.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    mine = generate.shard(everything, args.shard_index, args.shard_count)
+    if args.max_scenarios:
+        mine = mine[: args.max_scenarios]
+    output = Path(args.output_dir)
+    output.mkdir(parents=True, exist_ok=True)
+    results = []
+    for start in range(0, len(mine), BATCH):
+        batch = [scenario.to_json() for scenario in mine[start:start + BATCH]]
+        done, failure = _child(batch, output / f"batch-{start:05d}.json", args.control_every,
+                               SCENARIO_TIMEOUT * len(batch))
```
[the remark this answers is not in the document]
Comment (hhenson) [reply tools/recovery/cli.py]: Fixed in `915f1982a`. The child already rewrites its result file after every scenario, so the supervisor now watches that file and kills the child after **one** scenario's silence (plus a startup allowance for the first), instead of waiting out twenty scenarios' worth. The unreported rest are then run singly as before, so the culprit is named. Tested with a child that reports one result and hangs: the result is kept, the failure says `hung`, and it returns in seconds.

File: tests/cpp/test_recovery_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    void require_flat(const std::vector<Row> &rows, const char *label)
+    {
+        const auto per_key = [](double ms, std::size_t keys) { return ms * 1000.0 / static_cast<double>(keys); };
+        const auto &small = rows.front();
+        const auto &large = rows.back();
+        INFO(label);
+        if (small.save_ms > 2.0)
+            CHECK(per_key(large.save_ms, large.keys) < 3.0 * per_key(small.save_ms, small.keys) + 1.0);
+        if (small.restore_ms > 2.0)
+            CHECK(per_key(large.restore_ms, large.keys) < 3.0 * per_key(small.restore_ms, small.keys) + 1.0);
```
[the remark this answers is not in the document]
Comment (hhenson) [reply tests/cpp/test_recovery_benchmark.cpp]: Fixed in `915f1982a`. The guard was on the wrong measurement: a quadratic is exactly the case where the small run stays in the noise and the large one does not. It now skips only when the **large** run is below the noise floor, and a small run below the floor is read as "at most the floor" -- the most generous reading the clock allows -- rather than disabling the check. The +1 µs/key slack is gone with it.

**Comment by hhenson** [comment]
Full Python validation of `915f1982a` on a private Linux validation host, **real install** (core wheel, then the persistence extension built against its SDK): the whole `python/tests` suite **and** `extensions/persistence/python/tests` -- **4,593 passed, 9 skipped, 1 xfailed, 0 failed**. The xfail is the strict pin of the key-order defect. That covers the campaign sample, the supervisor test, the family-signature negative controls, and the architecture ratchets (green here because the one-line `WorkerPool::stop` fix from #998 is carried on this branch as well; the change is identical, so whichever merges first, the other merges cleanly).

🤖 Generated with [Claude Code](https://claude.com/claude-code)
